pdm/iostream.py

```diff
@@ -62,8 +58,10 @@ def __init__(self, verbosity: int = NORMAL, disable_colors: bool = False) -> Non
         logger.addHandler(logging.NullHandler())
         self.logger = logger
 
-        for color in COLORS:
-            setattr(self, color, functools.partial(self._style, fg=color))
+        self.green = functools.partial(self._style, fg="green")
+        self.cyan = functools.partial(self._style, fg="cyan")
+        self.yellow = functools.partial(self._style, fg="yellow")
+        self.red = functools.partial(self._style, fg="red")
```

Comment:
You sure? It has a call to `self`, so I'm not sure it makes sense as a class attribute.

In practice it won't make any difference because you instantiate a singleton and then pass that around. But having it as a class attribute makes it seem that `IoStream.cyan(...)` would work, when it wont

Reply:
I take that back. `self._style` should be a static method anyway, which would mean that the colours can be static attributes too.

Will update.

Reply:
> You sure? It has a call to self, so I'm not sure it makes sense as a class attribute

Yes, it's doable:

```python
class IOStream:
    ...
    def _style(...):
        ...
    green = functools.partial(_style, fg="green")
    red = functools.partial(_style, fg="red)
```

And I would prefer to keep them as instance methods, so that we can implement a customized IOStream object to replace it old one, in cases like testing.

Reply:
as per your comment i'v made the `_style` method a free function under the class namespace, and used `functools.partial` to set the class attributes. That's certainly a lot less verbose than writing out the methods manually.

I've defined the `bold` method in the same way as the colours. no reason to treat that one specially is there?

i've not changed any of the class attributes to instance attributes yet. Can you explain what you're after there? Which methods should be class attributes, and which should be instance attributes? If they're class attributes instead of instance attributes, why does that stop you writing a custom `IOSream` object with *different* class attributes for testing?

Reply:
actually it turns out you can't do this.

- if you define the `_style` function as an instance method mypy will complain that you haven't used `self`, so it should be a class method or a free function
- if you make it a class method (using the `@classmethod` decorator) then it's not callable from inside `functools.partial` because you need the class object to call it.
- This makes most sense as a free function, but then it's harder to override for testing.

so that leaves my original implementation where each of the 'styles' were implemented as class functions. This way no linter will complain, and the methods can be overridden for testing. It's a little verbose though, admittedly.

Reply:
> if you define the _style function as an instance method mypy will complain that you haven't used self, so it should be a class method or a free function

So you should define `supports_ansi` as an instance attribute(It's ok to not be a cached property) and reference `self.supports_ansi` from inside `_style()` then you can make `_style` a normal instance method without any complaints.

Reply:
> > if you define the _style function as an instance method mypy will complain that you haven't used self, so it should be a class method or a free function
> 
> So you should define `supports_ansi` as an instance attribute(It's ok to not be a cached property) and reference `self.supports_ansi` from inside `_style()` then you can make `_style` a normal instance method without any complaints.

same problem. `support_ansi` will have an used `self` parameter which linters will complain about. Also you can't then define the colour attributes in the class namespace since they need `self` in order to call `_style`.

The problem stems from the fact that almost all of the methods on `IoStream` are either stateless, or make use of a cached global state. The `IoStream` class is really only being used as a namespace

Reply:
> same problem. support_ansi will have an unused self parameter

Leave` _support_ansi()` as global func or staticmethod, some reasons why I insist on instance methods. `self.supports_ansi` is the state of `_style`.

1. Better to change instance attributes instead of class attributes in run time. It is natural to do so:
    ```python
    stream.supports_ansi = False
    # Do something else
    ```
2.  For the above reason, `_style()` also becomes an instance method.

Reply:
@frostming i've changed the methods/attributes *back* to instance methods.

note that this is where we started at the beginning of this pull request ;)

Reply:
You may not get me right, what i proposed is as following
```python
class IOStream:
    ...
    def _style(self, ...):
        ...
    green = functools.partial(_style, fg="green")
    red = functools.partial(_style, fg="red)
```

Binding methods in initializing looks a bit awkward to me.

Reply:
I don't understand what you mean? If you don't bind it in the class scope, instead of the `__init__` method, than it's not an instance attribute, it's a class attribute.
Can you explain what you mean? I'm happy to arrange this any way you like.

Reply:
For methods, as long as they are not decorated by classmethod or staticmethod, they are regular bound methods which accept `self` as the first argument and they are defined in the class scope. I might have taken the wrong word 'instance methods' for them, they are indeed called "bound methods"

Reply:
maybe i'm just being thick...
is this what you mean?

```python
class IOStream:
    NORMAL = 0
    DETAIL = 1
    DEBUG = 2

    def _style(self, text: str, *args, **kwargs) -> str:
        if self.supports_ansi:
            return click.style(text, *args, **kwargs)
        return text

    green = functools.partial(_style, fg="green")
    cyan = functools.partial(_style, fg="cyan")
    yellow = functools.partial(_style, fg="yellow")
    red = functools.partial(_style, fg="red")
    bold = functools.partial(_style, bold=True)

    def __init__(self, verbosity: int = NORMAL, disable_colors: bool = False) -> None:
        self.verbosity = verbosity
        self._indent = ""
        logger = logging.getLogger(__name__)
        logger.setLevel(logging.DEBUG)
        logger.addHandler(logging.NullHandler())
        self.logger = logger
        self.supports_ansi = _supports_ansi()
```

because this doesn't work.

because the methods ('green', 'cyan', etc.) are bound as class attributes, they're called without the `self` argument that `_style` requires.

<details>
<summary>pytest output</summary>

```
================================================================================= short test summary info =================================================================================
FAILED tests/test_project.py::test_global_project - TypeError: _style() missing 1 required positional argument: 'text'
FAILED tests/cli/test_actions.py::test_list_packages - TypeError: _style() missing 1 required positional argument: 'text'
FAILED tests/cli/test_cli.py::test_help_option - assert 'PDM - Python Development Master' in ''
FAILED tests/cli/test_cli.py::test_uncaught_error - assert '[RuntimeError]: test error' in ''
FAILED tests/cli/test_cli.py::test_show_package_on_pypi - assert 1 == 0
FAILED tests/cli/test_run.py::test_run_command_not_found - assert "Command 'foobar' is not found on your PATH." in ''
FAILED tests/models/test_candidates.py::test_expand_project_root_in_url[demo @ file:///${PROJECT_ROOT}/tests/fixtures/artifacts/demo-0.0.1-py2.py3-none-any.whl] - zipfile.BadZipFile: F...
ERROR tests/test_formats.py::test_convert_pipfile - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/test_formats.py::test_convert_requirements_file[False] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/test_formats.py::test_convert_requirements_file[True] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/test_formats.py::test_convert_requirements_file_without_name - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/test_formats.py::test_convert_poetry - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/test_formats.py::test_convert_flit - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/test_formats.py::test_convert_legacy_format - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/test_formats.py::test_import_requirements_with_section - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/test_plugin.py::test_plugin_new_command - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/test_plugin.py::test_plugin_replace_command - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/test_plugin.py::test_load_multiple_plugings - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/test_project.py::test_project_python_with_pyenv_support - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/test_project.py::test_project_config_items - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/test_project.py::test_project_config_set_invalid_key - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/test_project.py::test_project_sources_overriding - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/test_project.py::test_project_use_venv - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/test_project.py::test_project_with_combined_extras - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/test_project.py::test_project_packages_path - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/test_project.py::test_project_auto_detect_venv - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/test_project.py::test_ignore_saved_python - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_sync_only_different - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_sync_in_sequential_mode - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_sync_no_lockfile - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_sync_clean_packages - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_sync_dry_run - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_add_package[False] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_add_package[True] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_add_package_to_custom_package - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_add_editable_package[False] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_add_editable_package[True] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_add_remote_package_url[False] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_add_remote_package_url[True] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_remove_both_normal_and_editable_packages[False] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_remove_both_normal_and_editable_packages[True] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_add_no_install - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_add_package_save_exact - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_add_package_save_wildcard - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_add_package_update_reuse - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_add_package_update_eager - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_update_all_packages - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_update_specified_packages - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_update_specified_packages_eager_mode - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_remove_package[False] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_remove_package[True] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_remove_package_no_sync - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_remove_package_not_exist - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_remove_package_exist_in_multi_section - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_add_remove_no_package - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_update_with_package_and_sections_argument - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_add_package_with_mismatch_marker - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_add_dependency_from_multiple_parents - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_lock_dependencies - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_project_no_init_error - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_list_dependency_graph - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_list_dependency_graph_with_circular - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_list_reverse_without_graph_flag - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_list_reverse_dependency_graph - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_update_unconstrained_without_packages - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_update_ignore_constraints - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_init_validate_python_requires - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_actions.py::test_editable_package_override_non_editable - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_build.py::test_build_single_module - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_build.py::test_build_single_module_with_readme - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_build.py::test_build_package - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_build.py::test_build_src_package - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_build.py::test_build_package_include - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_build.py::test_build_src_package_by_include - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_build.py::test_build_legacy_package - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_lock_command - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_install_command - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_sync_command - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_update_command - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_remove_command - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_add_command - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_build_command - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_list_command - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_info_command - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_deprecate_global_project - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_global_project_other_location - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_use_command - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_use_python_by_version - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_install_with_lockfile - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_init_command - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_init_command_library - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_config_command - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_config_get_command - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_config_set_command - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_config_env_var_shadowing - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_config_project_global_precedence - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_cache_clear_command - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_import_other_format_file[requirements.txt] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_import_other_format_file[Pipfile] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_import_other_format_file[pyproject-poetry.toml] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_import_other_format_file[projects/flit-demo/pyproject.toml] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_search_package - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_export_to_requirements_txt - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_cli.py::test_lock_legacy_project - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_run.py::test_pep582_launcher_for_python_interpreter - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_run.py::test_run_cmd_script - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_run.py::test_run_cmd_script_with_array - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_run.py::test_run_shell_script - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_run.py::test_run_call_script - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_run.py::test_run_script_with_extra_args - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_run.py::test_run_expand_env_vars - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_run.py::test_run_script_with_env_defined - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_run.py::test_run_script_with_dotenv_file - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_run.py::test_run_script_override_global_env - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_run.py::test_run_show_list_of_scripts - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/cli/test_run.py::test_run_with_another_project_root - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/models/test_candidates.py::test_parse_local_directory_metadata[False] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/models/test_candidates.py::test_parse_local_directory_metadata[True] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/models/test_candidates.py::test_parse_vcs_metadata[False] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/models/test_candidates.py::test_parse_vcs_metadata[True] - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/models/test_candidates.py::test_parse_artifact_metadata[/home/dan/Code/python/pdm/tests/fixtures/artifacts/demo-0.0.1.tar.gz] - TypeError: _style() missing 1 required posit...
ERROR tests/models/test_candidates.py::test_parse_artifact_metadata[/home/dan/Code/python/pdm/tests/fixtures/artifacts/demo-0.0.1-py2.py3-none-any.whl] - TypeError: _style() missing 1 ...
ERROR tests/models/test_candidates.py::test_parse_metadata_with_extras - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/models/test_candidates.py::test_parse_remote_link_metadata - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/models/test_candidates.py::test_extras_warning - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/models/test_candidates.py::test_parse_abnormal_specifiers - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/models/test_candidates.py::test_parse_project_file_on_build_error - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/models/test_candidates.py::test_parse_project_file_on_build_error_no_dep - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/resolver/test_resolve.py::test_resolve_named_requirement - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/resolver/test_resolve.py::test_resolve_requires_python - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/resolver/test_resolve.py::test_resolve_allow_prereleases - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/resolver/test_resolve.py::test_resolve_with_extras - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/resolver/test_resolve.py::test_resolve_local_artifacts[/home/dan/Code/python/pdm/tests/fixtures/artifacts/demo-0.0.1.tar.gz] - TypeError: _style() missing 1 required positi...
ERROR tests/resolver/test_resolve.py::test_resolve_local_artifacts[/home/dan/Code/python/pdm/tests/fixtures/artifacts/demo-0.0.1-py2.py3-none-any.whl] - TypeError: _style() missing 1 r...
ERROR tests/resolver/test_resolve.py::test_resolve_vcs_and_local_requirements[False-/home/dan/Code/python/pdm/tests/fixtures/projects/demo] - TypeError: _style() missing 1 required pos...
ERROR tests/resolver/test_resolve.py::test_resolve_vcs_and_local_requirements[False-git+https://github.com/test-root/demo.git#egg=demo] - TypeError: _style() missing 1 required positio...
ERROR tests/resolver/test_resolve.py::test_resolve_vcs_and_local_requirements[True-/home/dan/Code/python/pdm/tests/fixtures/projects/demo] - TypeError: _style() missing 1 required posi...
ERROR tests/resolver/test_resolve.py::test_resolve_vcs_and_local_requirements[True-git+https://github.com/test-root/demo.git#egg=demo] - TypeError: _style() missing 1 required position...
ERROR tests/resolver/test_resolve.py::test_resolve_local_and_named_requirement - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/resolver/test_resolve.py::test_resolving_auto_avoid_conflicts - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/resolver/test_resolve.py::test_resolve_conflicting_dependencies - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/resolver/test_resolve.py::test_resolve_no_available_versions - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/resolver/test_resolve.py::test_exclude_incompatible_requirements - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/resolver/test_resolve.py::test_union_markers_from_different_parents - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/resolver/test_resolve.py::test_requirements_from_different_sections - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/resolver/test_resolve.py::test_resolve_two_extras_from_the_same_package - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/resolver/test_resolve.py::test_resolve_package_with_dummy_upbound - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/resolver/test_resolve.py::test_resolve_dependency_with_extra_marker - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/resolver/test_resolve.py::test_resolve_parent_from_multiple_sources - TypeError: _style() missing 1 required positional argument: 'text'
ERROR tests/resolver/test_resolve.py::test_resolve_circular_dependencies - TypeError: _style() missing 1 required positional argument: 'text'
```

</summary>

i'm sorry if i'm misunderstanding

Reply:
i'm sorry if i'm misunderstanding!

Reply:
Sorry my bad, I thought it would work, will merge shortly, thanks for the patience.

Also, I have come up with some refactor ideas to avoid usage of the `stream` singleton completely.  Thank you again.